libsolidity/AST.cpp

```diff
@@ -305,8 +306,12 @@ TypePointer FunctionDefinition::getType(ContractDefinition const*) const
 void FunctionDefinition::checkTypeRequirements()
 {
 	for (ASTPointer<VariableDeclaration> const& var: getParameters() + getReturnParameters())
+	{
 		if (!var->getType()->canLiveOutsideStorage())
 			BOOST_THROW_EXCEPTION(var->createTypeError("Type is required to live outside storage."));
+		if (!(var->getType()->externalType()) && getVisibility() >= Visibility::Public)
```

Comment:
You might reverse the order here so that we do not need to create the external type if it is not necessary.